### PR TITLE
feat(genai-util): add GenAI metric helpers and histograms

### PR DESCRIPTION
**Pull request by psx95**
## Which problem is this PR solving?

Re: #3681

Generative AI instrumentations need to record standardized OpenTelemetry metrics such as operation duration, token usage (input and output tokens), and time-to-first-chunk/time-to-first-token with explicit bucket boundaries defined in the GenAI semantic conventions.
This PR introduces shared metric creation and recording helper functions for `@opentelemetry/genai-util`

## Short description of the changes

- **Standard Histogram Creation**: Added helper functions with standard explicit bucket boundaries:
  - `createDurationHistogram` (`gen_ai.client.operation.duration`)
  - `createTokenUsageHistogram` (`gen_ai.client.token.usage`)
  - `createTimeToFirstChunkHistogram` (`gen_ai.client.operation.time_to_first_chunk`)
  - `createTimePerOutputChunkHistogram` (`gen_ai.client.operation.time_per_output_chunk`)
- **Unit Tests**: Added test suite covering bucket boundaries, histogram creation, and metric recording behavior with valid and edge-case inputs (`test/metrics.test.ts`).

## Important Info
 - This is the **second PR** in a series that attempt to breakdown PR #3677. The files in this PR are taken from PR #3677 and did not require any modifications for the package to compile. 
 - Based off:  #3709

### CODE REVIEW

File: packages/genai-util/src/metrics.ts [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+ *
+ * @experimental
+ */
+export interface MetricCreationOptions {
+  /** Metric instrument name override. */
+  name?: string;
+  /** Metric description override. */
+  description?: string;
+  /** Metric unit override. */
+  unit?: string;
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts]: why not histogram boundaries too ?

File: packages/genai-util/src/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+/**
+ * Create standard `gen_ai.client.operation.duration` histogram.
+ *
+ * @experimental
+ */
+export function createDurationHistogram(
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts]: this won't survive multiple metrics and the fallback to METRIC_GEN_AI_CLIENT_OPERATION_DURATION does not seem right. we should rather have a helper per individual metric and we should generate them, otherwise we'll spend all the time keeping this up to date. IF you want to have a hand-written version, consider writing in the way that it'd be generated later (boilerplace and repetitiveness is better than ad-hoc logic)

File: packages/genai-util/src/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    });
+  }
+}
+
+/**
+ * Record time to first chunk metric.
+ *
+ * @experimental
+ */
+export function recordTimeToFirstChunk(
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts]: not sure if this method helps with anything, why expose it instead of exposing `_recordDuration` directly?

File: packages/genai-util/src/metrics.ts (hunk 2 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: IIUC, you're recommending to bring this more in-line with [Python's implementation](https://github.com/open-telemetry/opentelemetry-python-genai/blob/59e6efe6c0be45a9da78f8f3c21c5d6ac4488ed1/util/opentelemetry-util-genai/src/opentelemetry/util/genai/instruments.py#L42), where we don't allow users to pass optional parameters to the functions. 

I did it this way since I thought - it offered more flexibility, but I think we can be opinionated here - so I will change it to only accept `meter` as the argument. 

Could you explain what you meant by: 
> this won't survive multiple metrics

File: packages/genai-util/src/metrics.ts (hunk 3 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: This was just syntactic sugar in a way, but I can remove it.

File: packages/genai-util/src/metrics.ts (hunk 1 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: I think this will be moot now since we moved to opinionated methods where the users are not allowed to configure the histogram being created.

We can remove this type.

File: packages/genai-util/src/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+import type { Attributes, Histogram, Meter } from '@opentelemetry/api';
+import {
+  ATTR_GEN_AI_TOKEN_TYPE,
+  GEN_AI_TOKEN_TYPE_VALUE_INPUT,
+  GEN_AI_TOKEN_TYPE_VALUE_OUTPUT,
+  METRIC_GEN_AI_CLIENT_OPERATION_DURATION,
+  METRIC_GEN_AI_CLIENT_OPERATION_TIME_TO_FIRST_CHUNK,
+  METRIC_GEN_AI_CLIENT_TOKEN_USAGE,
+  METRIC_GEN_AI_SERVER_TIME_TO_FIRST_TOKEN,
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts]: there is no need for this metric here - it's for inference servers, we can always add it later if someone will need it

File: packages/genai-util/src/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  16777216, 67108864,
+];
+
+/**
+ * Standard explicit bucket boundaries for GenAI client time to first chunk (in seconds).
+ *
+ * @experimental
+ */
+export const GENAI_TIME_TO_FIRST_CHUNK_BUCKETS = [
+  0.001, 0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10, 30,
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts]: Per semconv, `gen_ai.client.operation.time_to_first_chunk` should use the same explicit bucket boundaries as operation duration: `[0.01, 0.02, 0.04, 0.08, 0.16, 0.32, 0.64, 1.28, 2.56, 5.12, 10.24, 20.48, 40.96, 81.92]`.

File: packages/genai-util/src/metrics.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    {
+      description:
+        'Time to receive the first chunk, measured from when the client issues the generation request to when the first chunk is received in the response stream.',
+      unit: 's',
+      advice: {
+        explicitBucketBoundaries: GENAI_TIME_TO_FIRST_CHUNK_BUCKETS,
+      },
+    }
+  );
+}
```
Comment (lmolkova) [inline packages/genai-util/src/metrics.ts:111]: Consider adding `createTimePerOutputChunkHistogram` and `GENAI_TIME_PER_OUTPUT_CHUNK_BUCKETS` for `gen_ai.client.operation.time_per_output_chunk` (defined in `semconv.ts`).

File: packages/genai-util/src/semconv.ts [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+ */
+export const ATTR_GEN_AI_OPERATION_NAME = 'gen_ai.operation.name' as const;
+
+/**
+ * The Generative AI provider as identified by the client or server instrumentation.
+ *
+ * @example openai
+ * @example anthropic
+ * @example aws.bedrock
+ * @example google_genai
```
Comment (lmolkova) [inline packages/genai-util/src/semconv.ts]: `google_genai` is not a valid provider name in semconv, should `gcp.gen_ai` or `gcp.gemini` 

https://github.com/verdie-g/semantic-conventions-genai/blob/ab523d099aa16bd01e35210c6f1766d2bc1845e4/docs/registry/attributes/gen-ai.md?plain=1#L437-L439

File: packages/genai-util/src/metrics.ts (hunk 4 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Ok, I will leave it out for this PR.

File: packages/genai-util/src/metrics.ts (hunk 6 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:111]: Sure, I will add that.

File: packages/genai-util/src/semconv.ts (hunk 7 above)
Comment (psx95) [reply packages/genai-util/src/semconv.ts]: Good catch, thank you!

File: packages/genai-util/src/metrics.ts (hunk 5 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Good catch!

File: packages/genai-util/src/semconv.ts (hunk 7 above)
Comment (psx95) [reply packages/genai-util/src/semconv.ts]: Addressed in [4444cc6](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/4444cc680e521ea860de63a76df4f924528c9692).

File: packages/genai-util/src/metrics.ts (hunk 5 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Addressed in [4444cc6](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/4444cc680e521ea860de63a76df4f924528c9692).

File: packages/genai-util/src/metrics.ts (hunk 4 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Addressed in [4444cc6](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/4444cc680e521ea860de63a76df4f924528c9692).

File: packages/genai-util/src/metrics.ts (hunk 6 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:111]: Addressed in [4444cc6](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/4444cc680e521ea860de63a76df4f924528c9692).

File: packages/genai-util/src/metrics.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  });
+}
+
+/**
+ * Create standard `gen_ai.client.token.usage` histogram.
+ *
+ * @experimental
+ */
+export function createTokenUsageHistogram(meter: Meter): Histogram {
+  return meter.createHistogram(METRIC_GEN_AI_CLIENT_TOKEN_USAGE, {
```
Comment (hectorhdzg) [inline packages/genai-util/src/metrics.ts:84]: Should this set `valueType: ValueType.INT`? Semconv defines token usage as an integer metric, but histograms default to `DOUBLE` when this is omitted.

**Comment by hectorhdzg** [comment]
Looks like some of these helpers were removed in later commits. Could you update the PR description to reflect what’s currently included?

**Comment by psx95** [comment]
> Looks like some of these helpers were removed in later commits. Could you update the PR description to reflect what’s currently included?

Updated the PR description. Thanks!

File: packages/genai-util/src/metrics.ts (hunk 8 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:84]: Agreed, good catch! I overrode the `valueType` in [bf90bed](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/bf90bede591b29f03ecff22a1683fd4e1298c6b5).

**Review by JacksonWeber — APPROVED** [review]
LGTM

File: packages/genai-util/src/metrics.ts [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  GEN_AI_TOKEN_TYPE_VALUE_INPUT,
+  GEN_AI_TOKEN_TYPE_VALUE_OUTPUT,
+  METRIC_GEN_AI_CLIENT_OPERATION_DURATION,
+  METRIC_GEN_AI_CLIENT_OPERATION_TIME_PER_OUTPUT_CHUNK,
+  METRIC_GEN_AI_CLIENT_OPERATION_TIME_TO_FIRST_CHUNK,
+  METRIC_GEN_AI_CLIENT_TOKEN_USAGE,
+} from './semconv';
+import type { TokenUsage } from './types';
+
+/**
```
Comment (aabmass) [inline packages/genai-util/src/metrics.ts:23]: nit I don't think we need to export all of these constants

File: packages/genai-util/src/metrics.ts [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+export function recordOperationDuration(
+  histogram: Histogram | undefined,
+  durationSeconds: number,
+  attributes?: Attributes
+): void {
+  if (!histogram || durationSeconds < 0 || !isFinite(durationSeconds)) {
+    return;
+  }
+  histogram.record(durationSeconds, attributes);
+}
```
Comment (aabmass) [inline packages/genai-util/src/metrics.ts:146]: IMO this function isn't adding much value, can the caller just do `histogram.record(...)`?

File: packages/genai-util/src/metrics.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  if (typeof usage.inputTokens === 'number' && usage.inputTokens >= 0) {
+    histogram.record(usage.inputTokens, {
+      ...attributes,
+      [ATTR_GEN_AI_TOKEN_TYPE]: GEN_AI_TOKEN_TYPE_VALUE_INPUT,
+    });
+  }
+
+  if (typeof usage.outputTokens === 'number' && usage.outputTokens >= 0) {
```
Comment (aabmass) [inline packages/genai-util/src/metrics.ts:169]: I would have to double check what we do in python or define in semconv, but it feels weird to drop the zeros on the floor. Was there a reason for this?

File: packages/genai-util/src/metrics.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+ * @experimental
+ */
+export function recordTokenUsage(
+  histogram: Histogram | undefined,
+  usage: TokenUsage | undefined,
+  attributes?: Attributes
+): void {
+  if (!histogram || !usage) {
+    return;
+  }
```
Comment (aabmass) [inline packages/genai-util/src/metrics.ts:160]:
```suggestion
  histogram: Histogram,
  usage: TokenUsage,
  attributes?: Attributes
): void {
```

unless there is a good reason, can we just forbid undefined?

File: packages/genai-util/test/metrics.test.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  beforeEach(() => {
+    metricReader = new TestMetricReader();
+    meterProvider = new MeterProvider({ readers: [metricReader] });
+  });
+
+  afterEach(async () => {
+    await meterProvider.shutdown();
+  });
+
+  it('should define standard bucket boundaries', () => {
```
Comment (aabmass) [inline packages/genai-util/test/metrics.test.ts]: I guess this is why we export the constants, but not sure these tests provides much value. If you really wanted you could validate the buckets on the actual metric produced by the SDK

File: packages/genai-util/test/semconv.test.ts [hunk 14]
```diff
@@ -100,6 +102,14 @@ describe('GenAI Semantic Conventions', () => {
       METRIC_GEN_AI_CLIENT_TOKEN_USAGE,
       'gen_ai.client.token.usage'
     );
+    assert.strictEqual(
```
Comment (aabmass) [inline packages/genai-util/test/semconv.test.ts:105]: oop missed these before but testing the values of the constants is unnecessary imo

**Review by aabmass — CHANGES_REQUESTED** [review]

File: packages/genai-util/src/metrics.ts (hunk 11 above)
Comment (aabmass) [reply packages/genai-util/src/metrics.ts:169]: Oops missed the `>=`. But still it stands, I don't think we need to validate the value and drop it on the floor. The SDK should already handle this.

File: packages/genai-util/test/metrics.test.ts (hunk 13 above)
Comment (psx95) [reply packages/genai-util/test/metrics.test.ts]: Yeah, I will clean up some unnecessary tests in the PR. Thanks for the review! 

Addressed in [434c2e9](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/434c2e9ddcf9e1e3c0138bcffa43e758ca9aa34b).

File: packages/genai-util/src/metrics.ts (hunk 9 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:23]: Good point, they're used only within this file.

Addressed in [434c2e9](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/434c2e9ddcf9e1e3c0138bcffa43e758ca9aa34b).

File: packages/genai-util/src/metrics.ts (hunk 10 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:146]: Agreed, actually it is going to be removed in the next PR - https://github.com/psx95/opentelemetry-js-contrib/pull/2/changes#diff-bb9bbaf5f970e3932f8afc63a2c8c6851da1b5827a5f55719a0b6c6ba4c930f8L132. 

Technically it could be removed in this PR as well, but I realized this function is extraneous while working on the next one. Let me know if you think I should remove it here instead.

File: packages/genai-util/src/metrics.ts (hunk 12 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:160]: I can address this in the later PRs when this is actually used - in the core invocations. This function in its current form will be removed in the next PR.

File: packages/genai-util/src/metrics.ts (hunk 11 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts:169]: I'm not sure what is meant by 'Drop it to the floor'. Could you expand on this a little? 

The purpose of this is to only record token usage if its reported as a positive number.

I'm not sure if all SDKs handle it - I think its simple enough to check it here.

File: packages/genai-util/test/semconv.test.ts (hunk 14 above)
Comment (psx95) [reply packages/genai-util/test/semconv.test.ts:105]: I agree, but then `semconv.ts` has no testable logic. These tests keep the coverage numbers up which could go dramatically down as semconv.ts is huge. 

Technically, we could delete the entire `semconv.test.ts`. Let me know if you have any suggestions around this.

File: packages/genai-util/src/metrics.ts (hunk 11 above)
Comment (aabmass) [reply packages/genai-util/src/metrics.ts]: The SDK is intended to handle negatives ([spec](https://opentelemetry.io/docs/specs/otel/metrics/api/#:~:text=The%20value%20is%20expected%20to%20be%20non%2Dnegative.%20This%20API%20SHOULD%20be%20documented%20in%20a%20way%20to%20communicate%20to%20users%20that%20this%20value%20is%20expected%20to%20be%20non%2Dnegative.%20This%20API%20SHOULD%20NOT%20validate%20this%20value%2C%20that%20is%20left%20to%20implementations%20of%20the%20API.)). The JS SDK handles negatives and non finite values already https://github.com/open-telemetry/opentelemetry-js/blob/3d1d427a445d6d1b76b94785c3d0f4048e38cc69/packages/sdk-metrics/test/Instruments.test.ts#L283-L298

For the JS SDK, it also has some logging which is being bypassed here https://github.com/open-telemetry/opentelemetry-js/blob/3d1d427a445d6d1b76b94785c3d0f4048e38cc69/packages/sdk-metrics/src/Instruments.ts#L121-L126

File: packages/genai-util/src/metrics.ts (hunk 10 above)
Comment (aabmass) [reply packages/genai-util/src/metrics.ts]: > Let me know if you think I should remove it here instead.

SGTM

File: packages/genai-util/src/metrics.ts (hunk 12 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Removed this function in [391f19e](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/391f19ef3c4fe936cb4f63c95f5942036ecfe1d8).

File: packages/genai-util/src/metrics.ts (hunk 11 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Removed this function in [391f19e](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/391f19ef3c4fe936cb4f63c95f5942036ecfe1d8).

File: packages/genai-util/src/metrics.ts (hunk 10 above)
Comment (psx95) [reply packages/genai-util/src/metrics.ts]: Removed this function in [391f19e](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3712/commits/391f19ef3c4fe936cb4f63c95f5942036ecfe1d8).

**Review by hectorhdzg — APPROVED** [review]
LGTM, thanks for your contribution

File: packages/genai-util/test/semconv.test.ts (hunk 14 above)
Comment (aabmass) [reply packages/genai-util/test/semconv.test.ts:105]: OK can clean them up in a later PR

Comment (psx95) [reply packages/genai-util/test/semconv.test.ts:105]: Will do a cleanup after the remaining PRs are merged. Thanks!

**Review by lmolkova — APPROVED** [review]
LGTM!

Please update PR description

**Review by eternalcuriouslearner — APPROVED** [review]

**Review by aabmass — APPROVED** [review]
